Drivers: ARM: SysTick: investigate errors and apply fixes

ioannisg opened the pull request:
Several fixes/improvements in the SysTick timer driver, including
- correct calculation of elapsed cycles in the ISR
- avoiding starving the ISR if timeouts gets re-scheduled all the time
- reworking the way elapsed cycles are calculated

Added some documentation in functions and variables, too.

The fixes are closing several ARM timer-related bugs:

Fixes #18591 which is an instance of #20585
Fixes #20445 which is also an instance of #20585 
Also incorporates the SysTick modification related to #20939

### Mandatory Testing in Hardware

(Purpose is to get confidence on the driver changes.)

**All** platforms run with the **SysTick** timer driver.

| Cortex-M Platform | Test latest PR 633f18d4df89b4e6c06aa75605de721e947c9272 | kernel (76) | benchmark (5) | timer (1) | critical (1) | usermode (54)|
| :---         | :---: |  :---:      |    :---: |  :---: | :---: | :---: |      
| arduino_due (M3, 84MHz) | **yes** | :white_check_mark: | ✔️ | ✔️ | ✔️ | ➖ |
| atsamd21_xpro (M0+, 48MHz)   | **yes** | ✔️ | ✔️ |  ✔️  |  ✔️ | ✔️ |
| atsamr21_xpro (M0+, 48MHz)   | **yes** | ✔️ | ✔️ |  ✔️  |  ✔️ | ⚪️ |
| efr32mg_sltb004a (M4, 38.4 MHz) | **yes** | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| frdm_k64f (M4, 120MHz) | **yes** | ✔️ | ✔️ | ✔️ | ✔️  | ✔️ |
| lpcxpresso54114_m4 (M4, 48MHz ) | **yes** | ✔️ | ✔️ | ✔️ | ✔️ | ➖ |
| mimxrt1015_evk (M7, 500MHz) | **yes** | ✅ | ✔️ | ✔️ | ✔️ | ➖ |
| mimxrt1064_evk (M7, 600MHz) | **yes** |  ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| nRF52_pca10040 (M4, 64MHz)  | **yes** | ✔️ | ✔️ | ✔️ | ✔️ | ➖ |
| nRF5340_dk_nrf5340_app (M33, 128MHz) | **yes** | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| nRF9160_pca10090 (M33, 64MHz)  | **yes** | ✔️❔1 | ✔️ | ✔️ | ✔️ | ✔️ |
| nucleo_g071rb (M0+, 64MHz) | **yes** | ✅ | ✔️ | ✔️ | ✔️ | ➖ |
| nucleo_l476rg (M4, 80 MHz) |**yes**| ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
| sam_e70_xplained (M7, 300MHz) | **yes**| ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |

Legend

- ⚪️ tests not supported
- ➖ not tested
- ✔️tests passed
- :white_check_mark: selected tests (common, context, fifo, queue, shed, sleep, tickless, timer) passed
- :ballot_box_with_check: tests passed (sporadic failures observed)
- ❔1: timer/timer_api fails sporadically with zephyr_sdk. test passes with gnuarmemb toolchain
- :x: tests failed

ioannisg committed:
drivers: timer: SysTick: remove unnecessary masking

Unsupported bits of the Current Value Register
are read as zero, so we remove the redundant
ANDing with the max supported counter value.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
drivers: timer: SysTick: document internal function and variables

Add detailed documentation for the internal 'elapsed()'
function, as well as for the local counter variables used
in the SysTick driver.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
drivers: systick: fix calculation of absolute cycles count in ISR

The commit fixes the update of the absolute counter of HW cycles
in the SysTick ISR for TICKLESS mode.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@nashif , DNM to 1.14, until there's feedback that this is safe to go. Basically, I want to make sure this won't break anything.

@MaureenHelm could you run tests/kernel/critical on the fast NXP platforms and provide some feedback?
@andyross this is a second attempt to fix the issue with forward clock drift in SysTick. Instead of supressing events, now, we simply replace the `last_load `with actual `elapsed()` cycles in the ISR. So, even if the ISR is triggered immediately after the timeout setting, it won't "simulate" a complete SysTick cycle.

Note: this is what the nRF timer driver does, i.e. uses the elapsed() call in the ISR.

I added some documentation in the elapsed() function - @andyross let me know if you agree with my perception of what the function is meant to do.

MaureenHelm commented:
> @MaureenHelm could you run tests/kernel/critical on the fast NXP platforms and provide some feedback?

Test passes on mimxrt1020_evk, mimxrt1050_evk, and frdm_k64f.

andyross reviewed (approved):
Looks correct.

Note that while you're in there, a reasonable optimization to non-tickless mode would be to turn the ISR into a near-noop that just calls z_clock_announce(1).  The accounting on last_load et. al. is all degenerate if all we want is a regular clock interrupt.

nashif commented:
breaks a new test:

```

sam_e70_xplained          tests/kernel/timer/timer_monotonic/kernel.timer    FAILED: failed
	see: sanity-out/sam_e70_xplained/tests/kernel/timer/timer_monotonic/kernel.timer/handler.log

```

nashif reviewed (changes requested):
few tests still failing with this change

ioannisg commented:
> 
> 
> breaks a new test:
> 
> ```
> 
> sam_e70_xplained          tests/kernel/timer/timer_monotonic/kernel.timer    FAILED: failed
> 	see: sanity-out/sam_e70_xplained/tests/kernel/timer/timer_monotonic/kernel.timer/handler.log
> ```
> 
> in addition to
> 
> ```
> sam_e70_xplained          tests/kernel/critical/kernel.common                FAILED: timeout
> 	see: sanity-out/sam_e70_xplained/tests/kernel/critical/kernel.common/handler.log
> ```

Hm, @andyross might want to look at the first failure, because it is somehow related to #15219. 
It seems the test is passing if I still keep the `& COUNTER_MAX ` in `val = SysTick->VAL & COUNTER_MAX;`

The failure does not seem to be related to the actual "bug" we are hunting here (3rd commit)

ioannisg commented:
> 
> 
> few tests still failing with this change

@nashif the first fail you pointed at is a test fail; we need to improve the test to account for the fact that measurement might not be monotonic for the SysTick, due to wrap 

> 
> 
> Looks correct.
> 
> Note that while you're in there, a reasonable optimization to non-tickless mode would be to turn the ISR into a near-noop that just calls z_clock_announce(1). The accounting on last_load et. al. is all degenerate if all we want is a regular clock interrupt.

Good, I will follow your proposal, @andyross , and will update the PR

pabigot reviewed (approved):
I've compared this solution against #18651.

* The original problem with tests/kernel/fifo/fifo_timeout/prj_poll was reproduced on atsamd21_xpro on master d5633bc395b2bc8ce9f7685e8df9ead43deb20c3: it always fails
* I rebased this on master d5633bc395b2bc8ce9f7685e8df9ead43deb20c3 resolving the conflict (see https://github.com/pabigot/zephyr/commits/pr/15487)
* tests/kernel/fifo/fifo_timeout/prj_poll passes reliably on atsamd21_xpro

This PR is currently blocked due to sanitycheck failures on sam_e70_xplained.  Investigation shows

* tests/kernel/timer/timer_monotonic/kernel.timer now passes reliably on simulation and on hardware
* tests/kernel/critical/kernel.common passes reliably on simulation and on hardware

So IMO this is good.  Discussion at https://github.com/zephyrproject-rtos/zephyr/pull/18651#discussion_r317940005 preferred this PR for this specific fix.

Could we please re-visit this and get it merged?  @nashif

ioannisg commented:
Thanks @pabigot 
So i am going to resolve the conflicts here and try to address CI issues. I'll run the relevant tests on all my ARM boards (have more than 15 now) and report the findings . I really really want to have this solved for 2.1 release;)

ioannisg commented:
- All `kernel` and `benchmarks` tests pass on the following ARM platforms with SysTick timer:
  -  `frdm_k64f`
  - `lpcxpresso54114_m4`
  - `sam_e70_xplained`
  - `atsamr21_xpro`
  - `atsamd21_xpro` (@pabigot can you confirm this?)

- The following tests/tests-suites are passing on `mimxrt1015_evk` platform:
   - kernel: fifo, timer_api, sched
- The `critical` test-suite hangs on `mimxrt1015_evk`, but it also hangs in master branch.

I will report tomorrow the results on more platforms.

ioannisg commented:
@agansari could you do some additional testing on NXP platforms (other than the ones I've already tested)? Also, do you think we can bring here the fixes for the QEMU platforms (i.e. removing the workaround) you've done in #18651 ?

ioannisg committed:
tests: kernel: add kernel and sleep tags in tests

Adding kernel tag in tests/kernel/early_sleep and sleep tag
in tests/kernel/sleep. So both early_sleep and sleep suites
have the same tags.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

agansari commented:
> @agansari could you do some additional testing on NXP platforms (other than the ones I've already tested)? Also, do you think we can bring here the fixes for the QEMU platforms (i.e. removing the workaround) you've done in #18651 ?

Tested tests/kernel/ on:
 - mimxrt1050_evk
 - frdm_k64f
 - frdm_kw41z
They all passed

Some tests fail on mps2_an385, need to revisit #18651 and fix the issue @pabigot has with the last commit.

ioannisg commented:
@agansari could you please, also test` ./tests/benchmarks` tests? Likely include timing-related testing, as well! Thanks in advance :)

ioannisg commented:
>Some tests fail on mps2_an385, need to revisit #18651 and fix the issue @pabigot has with the last commit.

@agansari I think, by just applying 4926d5ce2f597c71a7c96f2bb2dbd1a19de1a2e2 from #18651 here, makes the kernel tests all pass on mps2 (an385 and an521). That's what I could see locally - let me know if you notice something else ;)

agansari commented:
> > Some tests fail on mps2_an385, need to revisit #18651 and fix the issue @pabigot has with the last commit.
> 
> @agansari I think, by just applying [4926d5c](https://github.com/zephyrproject-rtos/zephyr/commit/4926d5ce2f597c71a7c96f2bb2dbd1a19de1a2e2) from #18651 here, makes the kernel tests all pass on mps2 (an385 and an521). That's what I could see locally - let me know if you notice something else ;)

@ioannisg adding just [4926d5c](https://github.com/zephyrproject-rtos/zephyr/commit/4926d5ce2f597c71a7c96f2bb2dbd1a19de1a2e2) commit will not bring any actual tickless capability to QEMU targets as long as workarounds are not dropped, it may just improve the emulated ticker's accuracy.

ioannisg commented:
> l not bring any actual tickles

Right, I see. 
My proposal @agansari would be for you to resolve the QEMU issues right on top of this PR, **as long as** we have approved this driver fix and confirmed that any tests are only failing on emulated platforms.

I am going to run the timing tests on nRF platforms as well, using the SysTick as the timer driver. Just to get more confidence on the fix.

agansari commented:
> My proposal @agansari would be for you to resolve the QEMU issues right on top of this PR, as long as we have approved this driver fix and confirmed that any tests are only failing on emulated platforms.

@ioannisg can we merge #18651 before since it passes all tests? I think it just needs some renaming and review approval from @pabigot . I dropped the commit similar to this pull, local qemu tests and shippable pass.
Are you okay with this?

ioannisg commented:
> 
> 
> > My proposal @agansari would be for you to resolve the QEMU issues right on top of this PR, as long as we have approved this driver fix and confirmed that any tests are only failing on emulated platforms.
> 
> @ioannisg can we merge #18651 before since it passes all tests? I think it just needs some renaming and review approval from @pabigot . I dropped the commit similar to this pull, local qemu tests and shippable pass.
> Are you okay with this?

I am OK. Now this PR is only removing the workaround for QEMU in MPS2, no other code changes. If @pabigot is happy with the tolerance in the sched test, it could go in.

pabigot commented:
@ioannisg I believe your comment below was meant for #18651.  I'm happy with that PR, and once it goes in this one can be rebased on it and should then pass CI tests.

> > > My proposal @agansari would be for you to resolve the QEMU issues right on top of this PR, as long as we have approved this driver fix and confirmed that any tests are only failing on emulated platforms.
> > 
> > 
> > @ioannisg can we merge #18651 before since it passes all tests? I think it just needs some renaming and review approval from @pabigot . I dropped the commit similar to this pull, local qemu tests and shippable pass.
> > Are you okay with this?
> 
> I am OK. Now this PR is only removing the workaround for QEMU in MPS2, no other code changes. If @pabigot is happy with the tolerance in the sched test, it could go in.

ioannisg commented:
Rebased on top of #18651

ioannisg commented:
@nashif are you able to revisit this PR (possibly add some HW testing on top of what we've already done). Thanks!

psidhu commented:
@ioannisg I backported this into 1.14 where it solved an issue of mine. When this makes it in, can we put it into 1.14.2?

Tested this on a Cortex M4 (STM32L4R5+ series).

ioannisg commented:
> @ioannisg I backported this into 1.14 where it solved an issue of mine. When this makes it in, can we put it into 1.14.2?
> 
> Tested this on a Cortex M4 (STM32L4R5+ series).
> 
> EDIT: Actually, I say that, but it seems that when the cpu CycleCnt reaches 0x40000000, the CycleCnt stops incrementing. Have you seen this @ioannisg?

@psidhu could you give some details? which issue did it solve?

I've seen strange issues with the cycle_count, actually. Could you help us reproduce your last finding?

pabigot committed:
drivers: systick: avoid starving clock announcements

When setting a timeout measure the number of accumulated unannounced
ticks.  If this value exceeds half the 32-bit cycle counter range
force an announcement so the unannounced cycles are incorporated into
the system tick counter.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot reviewed (changes requested):
The top commit in https://github.com/pabigot/zephyr/commits/issue/annstarve will prevent time travel to the past in cases where tick announcements are starved as a result of too frequent rescheduling of the next timeout.

Either this patch, or a more general solution to #20585, should be identified and merged for 2.1 with the rest of what's in this PR.  IMO.

psidhu commented:
@ioannisg 

With regards to what it fixed in 1.14.1: For my application, I separately track the CPU cycle counts. This is in part because I need a high resolution clock reading that Zephyr today (or when I pinned to 1.14) doesn't necessarily provide (thus I use k_cycle_get_32()). When tracking this, I noticed a lot of clock slippage, especially when using tight timing loops. When I set CONFIG_SYS_CLOCK_TICKS_PER_SEC=10000000, the clock was slipping very noticeably. With this patch, that appeared to have gone away, but introduced a strange issue where it seems k_cycle_uptime_32() was not giving me updated times. I'll have to dig into this further to see exactly what the cause is. This was done using 0.10.0 SDK so further testing may be needed.

I've attempted to simplify the issue regarding the cycle_count issue on 2.0 using 0.10.3 SDK, but of course I'm having issues reproducing. However, with/without your patch, I'm still seeing the same slippage on 2.0. See the attached file, it's a "sample" program that shows the slippage. On my system, at about 100ms of uptime, there's a difference of 4-5ms which grows as it continues to run. I'm unsure if it's my loop, or something with k_uptime_get(). This is based on top of 0297756342.

[arm-systick.tar.gz](https://github.com/zephyrproject-rtos/zephyr/files/3833204/arm-systick.tar.gz)

pabigot commented:
@psidhu I think the problem is in your conversion from ticks to time.

Below is output from your program on mimxrt1060_evk with the following patch applied.  I suspect your system clock is not 10^N for some N, so a fractional component of cyc/ns is being truncated.  Using the new timer conversion works also.  I did not find the behavior to depend on `SYS_CLOCK_TICKS_PER_SEC`.

```
***** Booting Zephyr OS build zephyr-v2.0.0-1963-gc35a4690b854 *****
600000000 Hz cycles; cyc/ns = 1 = 1   ## NB: Should be 1.6666...
cy 11151ms ; up 18581 ms ; ca 18582 ms
cy 136423ms ; up 227369 ms ; ca 227370 ms
cy 139654ms ; up 232753 ms ; ca 232754 ms
cy 139755ms ; up 232921 ms ; ca 232922 ms
```

```
diff --git a/samples/hello_world/src/main.c b/samples/hello_world/src/main.c
index 48520382c2..e9de7a188f 100644
--- a/samples/hello_world/src/main.c
+++ b/samples/hello_world/src/main.c
@@ -47,6 +47,9 @@ void main(void)
 	k_timer_init(&idle_wake, idle_wake_timer_handler, NULL);
 	k_timer_start(&idle_wake, K_MSEC(1), K_MSEC(1));
 
+	printf("%u Hz cycles; cyc/ns = %u = %u\n",
+	       CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC, cycles_per_nsec,
+	       k_ns_to_cyc_ceil32(1));
 #ifdef USE_BUSY_THREAD
 	printf("USE_BUSY_THREAD Defined\n");
 	busy_tid1 = k_thread_create(
@@ -73,8 +76,8 @@ void main(void)
 		/* Convert to ms and print */
 		ns = total_system_cycles * cycles_per_nsec;
 		ms = ns / (NSEC_PER_USEC * USEC_PER_MSEC);
-		printf("\r%"PRIu32"ms %"PRIu32"ms",
-		       (uint32_t)ms, (uint32_t)k_uptime_ms);
+		printf("\rcy %"PRIu32"ms ; up %"PRIu32" ms ; ca %"PRIu32" ms",
+		       (uint32_t)ms, (uint32_t)k_uptime_ms, (u32_t)k_cyc_to_ms_ceil64(total_system_cycles));
 
 		/* Force idle */
 		k_cpu_idle();
```

psidhu commented:
@ioannisg Thanks. You're correct in that I wasn't doing the conversion correctly for the sample. I will attempt to dissect the issue with cycle_count in 1.14, but it seems that I don't see an actual issue in 2.0, or just can't reproduce it.

pabigot reviewed (changes requested):
The top commit of the updated https://github.com/pabigot/zephyr/commits/pr/15487 must be incorporated to avoid non-uniform uptime advances with nrf9160_pca10090 when using the systick timer.

ioannisg commented:
**All platforms run with the SysTick timer driver**

| ARM Platform | kernel (75) | benchmark (5) | timer (1) | critical (1) | userspace (~42)|
| :---         |     :---:      |    :---: |  :---: | :---: | :---: |      
| nRF5340_dk_nrf5340_cpuapp (M33, 128MHz)   | | |  |  | |
| nRF9160_pca10090 (M33, 64MHz)   | | |   |  | |
| nRF52840_pca10056 (M4, :  64MHz)   | | |  |  | |
| atsam_r21_xpro (M0+, 48MHz)   | ✔️ | ✔️ |  ✔️  |  ➖ | :white_circle: |
| sam_70_xplained (M7, 300MHz)   |  |  | ✔️ |  | |
| frdm_k64f (M4, 120MHz)   | ✔️ | ✔️ | ✔️ | ➖  | ✔️ |

Legend

- :white_circle: tests not supported
- ➖ tests did not run
- ✔️tests passed
- :x: tests failed

pabigot reviewed (approved):
Seems the volatile qualifiers on `cycle_count` and `announced_cycles` aren't necessary on `nrf9160_pca10090` with this combination of bits, so I've withdrawn my NAK, but there's something fragile going on there.

In my tests the only failures I get are on
* `sam_e70_explained` for `kernel.sched kernel.sched.multiq` which is due to the watchdog not being disabled.  Odd it doesn't happen on @ioannisg's board.

ioannisg commented:
> Seems the volatile qualifiers on `cycle_count` and `announced_cycles` aren't necessary on `nrf9160_pca10090` with this combination of bits, so I've withdrawn my NAK, but there's something fragile going on there.
> 
> In my tests the only failures I get are on
> 
> * `sam_e70_explained` for `kernel.sched kernel.sched.multiq` which is due to the watchdog not being disabled.  Odd it doesn't happen on @ioannisg's board.

Yes, I can confirm that the sched test-suite all passes on my sam_e70_xplained board.

pabigot commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -202,7 +203,13 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 		delay =
```
delay = CYC_PER_TICK * ceiling_fraction(delay, CYC_PER_TICK);

would be more succinct and clear, perhaps.

pabigot commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -202,7 +203,13 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+			last_load = delay;
+		} else {
+			/* If delay contains a value that cannot be programmed,
+			 * leave the previous value of last_load into SysTic.
```
Why leave this at whatever value it was before, rather than do:

    last_load = MAX_CYCLES;

to sleep for the longest allowable duration?  Also this is now inconsistent with the commit message which describes that behavior.

ioannisg replied:
I've been experimenting on this.
i think , if we reach a point that we provide trash to the systick LOAD register, I am not sure what the best policy should be.

pabigot replied:
But it's not "trash", it's simply a deadline that's too large to fit in the limited number of bits in `LOAD` and so has reached in steps.  The maximum tick-aligned value that does fit seems like the best policy; I don't know of another that would be defensible.  (Consider a desire to sleep for 2^25 cycles, but the last completed timeout was for 2^8 cycles.  The systick interrupt would have to fire 2^16 times to reduce the sleep to 2^24 which could be done in one step.)

pabigot commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -202,7 +203,13 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+			/* If delay contains a value that cannot be programmed,
+			 * leave the previous value of last_load into SysTic.
+			 */
+		}
```
Swap the next two lines?  (i.e. put the blank line after the last closing brace)

ioannisg committed:
drivers: timer: SysTick: enforce proper min & max SysTick LOAD values

Similar to what we do in other timer drivers, the maximum ticks
supplied in z_clock_set_timeout(..) needs to be MAX_TICKS at
maximum, when K_FOREVER is supplied as argument to the function.

In addition to that, the value we load onto the SysTick LOAD
register shall be truncated to MAX_CYCLES. This is required
to prevent loading a trash value to LOAD register, as only
the lowest 24 bits may be safely written.

Finally, we move the enforcement of the minimum delay to be
programmed on LOAD (i.e. MIN_DELAY) at the end step of the
calculation of the cycles-to-be-programmed. This prevents
from misscalculating the delay, as any required adjustment
is applied at the end, after the delay is rounded up to
the next tick boundary.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pabigot committed:
drivers: timer: SysTick: rework late overflow check

The previous solution depended on a magic number and was inefficient
(entered the second-wrap conditional even when a second wrap hadn't
been observed).  Replace with an algorithm that is deterministic.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Between @ioannisg and me we've tested the stuffing out of this thing, as summarized in the table in the commit description.

There is at least one intermittent failure of timer_api that occurs on nRF9160 (less than 1/10 tries, and platform doesn't normally use this driver).  We have ideas about the cause and will pursue that in the 2.2 development cycle.

There have been a couple additional changes to the overflow and timeout management if any existing reviewers want to revisit it.

But we both feel strongly that this provides significant improvements and should be merged for 2.1.

ioannisg commented:
I removed the HW Block label - i think this should go in for 2.1 release, unless anyone would like to do additional testing with other ARM Cortex-M platforms.

@andyross @nashif @galak @MaureenHelm ^^
